File: 424. Longest Repeating Character Replacement/Longest Repeating Character Replacement.cpp
#include <unordered_map>
#include <algorithm>

class Solution {
public:
    int characterReplacement(string s, int k) {
        int left = 0;
        unordered_map<char, int> hashMap;
        int result = 0;
        int maxCount = 0;
        
        for (int right = 0; right < s.length(); right++) {
            char value = s[right];
            hashMap[value] = hashMap[value] + 1;
            maxCount = max(maxCount, hashMap[value]);
            
            if ((right - left + 1 - maxCount) > k) {
                hashMap[s[left]] -= 1;
                left += 1;
            }
            
            result = max(result, right - left + 1);
        }
        
        return result;
    }
};

File: 424. Longest Repeating Character Replacement/Longest Repeating Character Replacement.js
/**
 * @param {string} s
 * @param {number} k
 * @return {number}
 */
var characterReplacement = function (s, k) {
  let left = 0;
  const hashMap = {};
  let result = 0;
  let maxCount = 0;

  for (let right = 0; right < s.length; right++) {
    const value = s[right];
    hashMap[value] = (hashMap[value] || 0) + 1;
    maxCount = Math.max(maxCount, hashMap[value]);

    if (right - left + 1 - maxCount > k) {
      hashMap[s[left]] -= 1;
      left += 1;
    }

    result = Math.max(result, right - left + 1);
  }

  return result;
};
